### DISCUSSION TITLE
Changing state variables used in JSBSim linearization model

### DISCUSSION THREAD

**Discussion by vrottolo** [JSBSim-Team/jsbsim#1083] [Q&A] [answered]
Hello everyone,

I'm currently working with JSBSim and I'm trying to understand how to modify the state variables used in the linearization model.

Currently, the state vector used for linearization is composed of

['Vt', 'Alpha', 'Theta', 'Q', 'Beta', 'Phi', 'P', 'Psi', 'R', 'Latitude', 'Longitude', 'Alt']. 

However, I would like to change it to the standard set of variables 

[u, v, w, p, q, r, X, Y, Z, phi, theta, psi].

Could anyone guide me on how to make this modification? Is it possible to customize the state variables for linearization in JSBSim?

Thank you in advance for your help!

**Comment by seanmcleod** [comment]
How are you planning on using the linearization model?

Depending on your use case, you do realize that all the states you're looking for can be derived from the existing state vector right?

**Comment by bcoconni** [comment, marked as answer]
@viniciusrottolo You could add new classes inheriting from `FGStateSpace::Component` for each of the variable you need. Similar to the class `FGStateSpace::Theta` below:
https://github.com/JSBSim-Team/jsbsim/blob/e26d03b2e0e94d49a76d7f13b4cdb9133d45dcbb/src/math/FGStateSpace.h#L402-L419
and modify the constructor `FGLinearization::FGLinearization` to use the components that you have created (replace each occurrence of `ss.x.add(new FGStateSpace::***)`)
https://github.com/JSBSim-Team/jsbsim/blob/e26d03b2e0e94d49a76d7f13b4cdb9133d45dcbb/src/initialization/FGLinearization.cpp#L28-L63

**Reply by vrottolo** [reply]
@bcoconni, can you help me with a question that might be stupid?

`simulation/do_simple_trim` changes the value of `fcs/pitch-trim-cmd-norm` and not `fcs/elevator-cmd-norm`, correct?

In this case, `DeCmd` is returning the value of `fcs/elevator-cmd-norm`, which is zero, right? 


https://github.com/JSBSim-Team/jsbsim/blob/e26d03b2e0e94d49a76d7f13b4cdb9133d45dcbb/src/initialization/FGLinearization.cpp#L61-L63

https://github.com/JSBSim-Team/jsbsim/blob/e26d03b2e0e94d49a76d7f13b4cdb9133d45dcbb/src/math/FGStateSpace.h#L610-L623

**Reply by bcoconni** [reply]
> `simulation/do_simple_trim` changes the value of `fcs/pitch-trim-cmd-norm` and not `fcs/elevator-cmd-norm`, correct?

That's correct: `fcs/elevator-cmd-norm` is left unmodified by `simulation/do_simple_trim` and pitch trim is used to trim $\dot{q}$.

https://github.com/JSBSim-Team/jsbsim/blob/b420d444e4709aabc20b5f8d8e69fa2890d4017d/src/initialization/FGTrim.cpp#L810
https://github.com/JSBSim-Team/jsbsim/blob/b420d444e4709aabc20b5f8d8e69fa2890d4017d/src/initialization/FGTrim.cpp#L821
https://github.com/JSBSim-Team/jsbsim/blob/b420d444e4709aabc20b5f8d8e69fa2890d4017d/src/initialization/FGTrim.cpp#L833
https://github.com/JSBSim-Team/jsbsim/blob/b420d444e4709aabc20b5f8d8e69fa2890d4017d/src/initialization/FGTrim.cpp#L842
> 
> In this case, `DeCmd` is returning the value of `fcs/elevator-cmd-norm`, which is zero, right?

Well, `fcs/elevator-cmd-norm` and `FGFCS::GetDeCmd` are returning the last value they have been set to. And this value depends on the FCS initialization for instance. It also depends on the previously simulated time steps if `simulation/do_simple_trim` is triggered at a time t > 0.
